### PR TITLE
Optimize RunLineHelper in BufferedInputStream

### PR DESCRIPTION
**Pull request by zhuzilin**
The old implementation of `ReadLineHelper` will append chars one at a time to the tstring result, which may cause multiply memcpy or reinitialization. This PR change from using
```c++
tstring& append(size_t n, char c)
```
to 
```c++
tstring& append(const char* str, size_t len)
```
to append multiple chars once.

This optimziation will help `TextlineDataset` which iteratively calls `ReadLineHelper`.

Thank you for your time on reviewing this PR.

### CODE REVIEW

File: tensorflow/core/lib/io/buffered_inputstream.cc [hunk 1]
```diff
@@ -60,25 +60,38 @@ Status BufferedInputStream::ReadLineHelper(StringType* result,
                                            bool include_eol) {
   result->clear();
   Status s;
+  size_t start_pos = pos_;
   while (true) {
     if (pos_ == limit_) {
+      if (start_pos != pos_) {
```
Comment (allenlavoie) [inline tensorflow/core/lib/io/buffered_inputstream.cc:66]: This gets repeated a few times and makes the logic pretty hard to follow.

Is it enough to call result->reserve when we fill the buffer? Probably OK if a bit of extra memory is reserved for the string.

**Comment by gbaned** [comment]
@zhuzilin  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

**Comment by zhuzilin** [comment]
@gbaned Sorry... I missed the email alert for this...
@allenlavoie 
I'm not sure how to use `result->reserve` to simplify the logic. The logic is:
1. If we read to the end of the buffer, we need to append the remaining to `result` and refill the buffer.
2. If the current character is `\n` we need to append to it and return.
3. skip `\r`.

I've removed all the `if (start_pos != pos_)` because I found that there is already an early return in `append` when the appending string is of zero length. Would this make the code a little better?

**Review by allenlavoie — APPROVED** [review]
